Include method in thumbnail media name

deepbluev7 opened the pull request:
This fixes an issue, where different crop methods overwrite one another.

Signed-off-by: Nicolas Werner <nicolas.werner@hotmail.de>

### Pull Request Checklist

<!-- Please read CONTRIBUTING.md before submitting your pull request -->

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

deepbluev7 commented:
I'm not sure, if this needs some kind of migration or similar, this is just the fix that worked for me, but I deleted my remote thumbnails manually...

deepbluev7 commented:
I think I also need to update the schema to include the method in the unique constraint, but I'm not really sure, how one would do that. It probably needs to look like mentioned in this comment: https://github.com/matrix-org/synapse/issues/2182#issuecomment-401297320

Constraint is for example here: https://github.com/matrix-org/synapse/blob/master/synapse/storage/data_stores/main/schema/full_schemas/54/full.sql.postgres#L1393

This may also solve #2182 then. I think I already dropped that constraint manually at some point, which is why I experienced this issue in the first place probably.

So some help on how one would approach that would be nice. Do I just add a delta file to the schema directory that drops the old constraint and creates a new one?

clokep commented:
> I think I also need to update the schema to include the method in the unique constraint, but I'm not really sure, how one would do that. It probably needs to look like mentioned in this comment: [#2182 (comment)](https://github.com/matrix-org/synapse/issues/2182#issuecomment-401297320)

Migrations are handled via deltas (some examples: https://github.com/matrix-org/synapse/tree/master/synapse/storage/data_stores/main/schema/delta/57). Unfortunately I don't seem to find any good documentation about how those are handled.

> So some help on how one would approach that would be nice. Do I just add a delta file to the schema directory that drops the old constraint and creates a new one?

I believe this is true, I think you'll need to create a new delta directory and add the appropriate sql in there. Also note whether a different query will need to be done for sqlite vs. postgresql.

deepbluev7 commented:
Thanks, @clokep , I'll take a look this weekend :D

richvdh commented:
You don't need to create a new directory unless the change you are making will be incompatible with older versions of synapse. Just stick a delta file in the latest directory (currently https://github.com/matrix-org/synapse/tree/develop/synapse/storage/data_stores/main/schema/delta/58). Number it so that it runs in the expected order.

There are a few words about how it works at https://github.com/matrix-org/synapse/blob/develop/synapse/storage/prepare_database.py#L102, though tbh that looks like it could do with a bit of an update to reflect the fact that there are now multiple "data stores".

deepbluev7 commented:
I hope the db migrations are correct, but SQL really isn't my forte :D

richvdh commented on synapse/rest/media/v1/filepath.py:
```diff
@@ -80,7 +80,7 @@ def remote_media_thumbnail_rel(
     ):
         top_level_type, sub_type = content_type.split("/")
-        file_name = "%i-%i-%s-%s" % (width, height, top_level_type, sub_type)
+        file_name = "%i-%i-%s-%s-%s" % (width, height, top_level_type, sub_type, method)
```
won't this mean that all existing thumbnails get broken and orphaned?

deepbluev7 replied:
They will get orphaned. I'm not sure how to clean them up in this case. If they get broken, I don't know. At least my synapse regenerates them, but I also have dynamic thumbnailing enabled.

What would be a reasonable way to clean them up? Most thumbnails potentially are already broken with dynamic thumbnailing, as we can't really be sure, with what method they were generated and if they aren't truncated.

deepbluev7 replied:
They will get orphaned. I'm not sure how to clean them up in this case. If they get broken, I don't know. At least my synapse regenerates them, but I also have dynamic thumbnailing enabled.

What would be a reasonable way to clean them up? Most thumbnails potentially are already broken with dynamic thumbnailing, as we can't really be sure, with what method they were generated and if they aren't truncated.

richvdh replied:
I agree that if you've got dynamic thumbnails enabled, then you have no real idea what is on the filesystem and you're best off removing them. (It looks like you're correct that the file will get recreated in this case). We should make sure not to orphan them, but that might be as simple as telling admins to delete the whole `remote_thumbnails` directory before upgrading, or give them a script to detect and delete files with the wrong name which they can run after the upgrade.

The problem is that most thumbnails are *not* broken when `dynamic_thumbnails` are disabled, so having them all orphaned (or even deleting them all so that they can be recreated is unacceptable).

A few options spring to mind:
1. record in the database somehow that these thumbnails are using the "legacy" filename
2. have the code fall back to the legacy filename if a file that the database says should exist is missing
3. try to figure out the likely method which was used for each thumbnail, and rename them accordingly.

I think option 2 is probably best?

deepbluev7 replied:
I think 2 is probably the easiest to implement and if marked correctly, it could probably be removed in the future, so I'll try to go for that. The rename in 3 would probably not be that much faster than thumbnailing images again as needed and would require a slow migration too. 1 would be doable, but would need another schema update in the future and I expect that table to be pretty big, so it may not be a good idea to add such baggage. :D

deepbluev7 replied:
Hm, I'm not so sure how to do any of this. I think I have an idea, how one could do method 3, if they didn't use dynamic thumbnailing (with dynamic you would need to wipe the cache anyway). You want to delete all your remote thumbnails if you had dynamic thumbnailing enabled, since you can't be sure, if your thumbnails are broken or not. If you didn't enable dynamic thumbnails, you can rename your remote thumbnails according to the methods in your homeserver.yaml, if they don't overlap. This takes a considerable disk overhead though and it may be easier to just delete them entirely and let them be recreated on demand, so that you don't need to rename a few million files.

Method 1 is pretty hard, if you want to handle all configurations correctly, since you can't register the "legacy" name for dynamic thumbnails without sometimes serving broken ones. It also now needs an additional db lookup to return the filename, which may lead to performance regressions.

Method 2 adds additional file system accesses for every cache miss and for legacy thumbnails, since you need to do 2 stats instead of 1.

So I probably prefer going with Method 3 and provide a script to rename the files or to just recommend to wipe the remote thumbnails to users. Otherwise I may need some help to implement 1 or 2.

I've already struggled quite a bit to understand how the background index update is supposed to work and I'm really not familiar enough with synapse/python to be confident to make such more complex changes.

richvdh replied:
> Method 2 adds additional file system accesses for every cache miss and for legacy thumbnails, since you need to do 2 stats instead of 1.

I don't think this is correct. If it's a cache miss, the media will not be in the database, so we won't even bother looking at the filesystem. For legacy thumbnails, it's true that we'll end up with two stats, but I think that's a price we can pay for now.

If it's a real problem, we can write a separate change later to work through the database and rename any files with the wrong name (or to rename them as we find them). But for now I think we can keep things simpler by checking both filenames.

deepbluev7 commented:
@richvdh Thank you for your review. I'll see if I can make the migration run in the background. About the orphaning I have no idea and would like some input on how to handle that. (Also sorry about the potential reply spam, had some issues with Github 500ing me...)

richvdh commented:
@deepbluev7 thanks very much for picking this up and working on it. As you can tell from the age of #2182, it's a very long-standing issue: I think lots of people will appreciate a fix!

deepbluev7 commented:
Well, it's scratch your own itch... I changed my client to display cropped images instead of scaled images and suddenly I had missing avatars, so I dug into it and stumbled on this :D

clokep commented:
@deepbluev7 The current packaging issue is unrelated to this PR (merging develop in should fix the issue). Let us know when this is ready for another review!

deepbluev7 commented:
@clokep Well, I'm currently stuck on how to prevent regenerating all remote thumbnails as mentioned in https://github.com/matrix-org/synapse/pull/7124#discussion_r422705424. If I figure out a solution on my own, I will let you know but I'd also appreciate some help on that topic. :D

Anyway, thanks for your ping. Always appreciated to know, that people care about what you are doing. :3

richvdh commented:
@deepbluev7 sorry for taking a while to get back to you on this. I've tried to respond to your question: do you think you can proceed now?

deepbluev7 commented:
Yeah, I think I can. Although it may take a few days to figure that out, I'll try if I can :3

clokep commented:
> (I need to figure out, where the file gets accessed, since I only changed the function, that returns the filename. I need to figure out, where that filename is used and implement a fallback there.)

According to GitHub there's only a single caller:

https://github.com/matrix-org/synapse/blob/737b4a936e35daaa839e7296d888041941546b47/synapse/rest/media/v1/media_storage.py#L214

deepbluev7 commented:
The issue is that that usage location is not actually, where the file is read. That just converts the path to a relative path. I need to modify the location where that path is read from though, so that if the read fails, it falls back to a different path. So I actually need to modify 2 locations:
https://github.com/matrix-org/synapse/blob/737b4a936e35daaa839e7296d888041941546b47/synapse/rest/media/v1/media_storage.py#L143 and https://github.com/matrix-org/synapse/blob/737b4a936e35daaa839e7296d888041941546b47/synapse/rest/media/v1/media_storage.py#L167 from what I can tell. :3

deepbluev7 commented:
I think this should work, but I still need to deploy it to my server for further testing.

richvdh commented:
@deepbluev7 really hoping you can find time to get back to this soon -  it would be great to see this fix landing!

deepbluev7 commented:
@richvdh  I sadly don't have much time to work on this atm. I did have time when I started this, but currently I'm busy with university, GSoC mentoring, work, Nheko etc. I'm not sure, if I can make time for it this month, but I'll see, what I can do. I should have some time in August though.

deepbluev7 commented on synapse/storage/data_stores/main/schema/delta/58/add_method_to_thumbnail_constraint.sql.postgres:
```diff
@@ -0,0 +1,31 @@
+-- drop old index
+INSERT INTO background_updates (update_name, progress_json, depends_on) VALUES
+  ('remote_media_repository_drop_index_wo_method', '{}', 'remote_media_repository_thumbnails_method_idx');
+
```
I don't quite understand, what this means. What is an ordering and where do I put it?

deepbluev7 commented:
I hope I addressed most of the comments. Since it has been a while, should I rebase this on current develop? And if I do that, do I need to move the delta files?

clokep commented:
@deepbluev7 Looks like this has some conflicts so we'll need to do something. [We prefer if you would merge develop into your branch (instead of a rebase)](https://github.com/matrix-org/synapse/blob/develop/docs/dev/git.md) since it makes it clearer what has changed.

I believe the conflicts are that we renamed `self.db` to `self.db_pool`. Hopefully those are straightforward to fix! 👍

deepbluev7 commented:
Thanks, i will do that. So I don't need to move the schemas?

clokep commented:
You should not need to move the schemas! Sorry missed that question. 😄

deepbluev7 commented:
Thank you, I guess we'll see how much that broke

clokep commented:
It doesn't seem that the python 3.5 failures are from this branch (I saw them another PR as well). I believe they're due to pypa/setuptools#2352.

deepbluev7 commented:
Yeah, I somehow doubt my changes could have broken everything on python 3.5 :D

richvdh reviewed (approved):
lgtm!

richvdh commented on changelog.d/7124.bugfix:
```diff
@@ -0,0 +1 @@
+Fix a bug in the media repository where remote thumbnails with the same size but different crop methods would overwrite each other.
```
```suggestion
Fix a bug in the media repository where remote thumbnails with the same size but different crop methods would overwrite each other. Contributed by @deepbluev7.
```

richvdh commented:
thanks again!

deepbluev7 commented:
Huh, that went pretty quick now. Thank you for all the help!
